### PR TITLE
new panos module that adds arbitrary service to the FW

### PR DESCRIPTION
**Pull request by ivanbojer**
##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
panos_service

##### ANSIBLE VERSION

```
ansible 2.1.0.0 
config file = 
configured module search path = Default w/o overrides
```

##### SUMMARY
New panos module that allows for admin to add arbitrary service definition to the firewall. Service definitions can be used in policies for enforcement.

### CODE REVIEW

File: lib/ansible/modules/network/panos/panos_service.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  - name: create SSH service
+    panos_service:
+      ip_address: "192.168.1.1"
+      password: "admin"
+      service_name: "service-tcp-22"
+      protocol: "tcp"
+      port: "22"
+'''
+
+RETURN = '''
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: return values do not match RETURN doc string

File: lib/ansible/modules/network/panos/panos_service.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+'''
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+
+from ansible.module_utils.basic import AnsibleModule
+
+try:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py:93]: Shared module?

File: lib/ansible/modules/network/panos/panos_service.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        port=dict(required=True),
+        source_port=dict(),
+        commit=dict(type='bool', default=True)
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+    if not HAS_LIB:
+        module.fail_json(msg='pan-python is required for this module')
+
+    ip_address = module.params["ip_address"]
+    if not ip_address:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: not needed

File: lib/ansible/modules/network/panos/panos_service.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        module.fail_json(msg="ip_address should be specified")
+    password = module.params["password"]
+    if not password:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: not needed

File: lib/ansible/modules/network/panos/panos_service.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    username = module.params['username']
+
+    xapi = pan.xapi.PanXapi(
+        hostname=ip_address,
+        api_username=username,
+        api_password=password
+    )
+
+    service_name = module.params['service_name']
+    if not service_name:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: not needed

File: lib/ansible/modules/network/panos/panos_service.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        module.fail_json(msg='service_name is required')
+    protocol = module.params['protocol']
+    if not protocol:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: not needed

File: lib/ansible/modules/network/panos/panos_service.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        module.fail_json(msg="protocol is required")
+    port = module.params['port']
+    if not port:
```
Comment (privateip) [inline lib/ansible/modules/network/panos/panos_service.py]: not needed

**Review by privateip — CHANGES_REQUESTED** [review]
consider adding unit tests

File: lib/ansible/modules/network/panos/panos_service.py (hunk 1 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: followed example from wakeonlan.py given that there are no return values.

File: lib/ansible/modules/network/panos/panos_service.py (hunk 2 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py:93]: no

File: lib/ansible/modules/network/panos/panos_service.py (hunk 3 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: ok

File: lib/ansible/modules/network/panos/panos_service.py (hunk 4 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: ok

File: lib/ansible/modules/network/panos/panos_service.py (hunk 5 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: ok

File: lib/ansible/modules/network/panos/panos_service.py (hunk 6 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: ok

File: lib/ansible/modules/network/panos/panos_service.py (hunk 7 above)
Comment (ivanbojer) [reply lib/ansible/modules/network/panos/panos_service.py]: ok

**Comment by ivanbojer** [comment]
Thank you for review. I also change the exception handling based on the previous comments.

**Comment by ivanbojer** [comment]
ready_for_review

**Pull request merged**